File: sdk/dotnet-guest/examples/streaming.cs
using Lsf.Guest;
using Raw = ServiceWorld.wit.Imports.latent.http.IStreamingImports;
namespace ServiceWorld.wit.Exports.tests.streamingHttp;
public class ApiExportsImpl : IApiExports
{
    public static ulong Run(uint which, string text, ulong handle)
    {
        using var scope = new Scope();
        var opened = Streaming.Open(new Raw.Request(Raw.Method.POST, text, [], 4, "text/plain", null, 1000));
        if (!opened.IsOk) return opened.AsErr.Tag == Raw.HttpError.Tags.PermissionDenied
            ? 10UL : throw new System.InvalidOperationException("unexpected streaming error");
        var upload = scope.Own(opened.AsOk);
        if (which == 1) return 1;
        _ = Streaming.Write(upload, [100, 97, 116, 97]).AsOk;
        var response = Streaming.Finish(upload).AsOk;
        var body = scope.Own(response.Body);
        if (which == 2) return 2;
        var chunk = scope.Own(Streaming.Read(body, 4).AsOk ?? throw new System.InvalidOperationException("missing chunk"));
        if (which == 3) {
            body.Dispose();
            return (ulong)Streaming.ChunkBytes(chunk).AsOk.Length;
        }
        var bytes = Streaming.ChunkBytes(chunk).AsOk;
        if (Streaming.Read(body, 4).AsOk is not null) throw new System.InvalidOperationException("extra chunk");
        _ = Streaming.Trailers(body).AsOk;
        var again = Streaming.Trailers(body);
        if (again.IsOk || again.AsErr.Tag != Raw.HttpError.Tags.InvalidState) throw new System.InvalidOperationException("trailers replayed");
        return (ulong)bytes.Length;
    }
}

File: sdk/java-guest/examples/streaming.java
package dev.latent.app;
import dev.latent.generated.Bindings;
import dev.latent.generated.Bindings.*;
import dev.latent.guest.*;
import java.util.List;

public final class Capsule implements Bindings.Exports {
    public Unsigned64 run(Long which, String text, Unsigned64 handle) {
        var opened = LatentHttpStreaming.open(new LatentHttpStreamingRequest(LatentHttpStreamingMethod.Post,
            text, List.of(), Option.some(new Unsigned64(4)), Option.some("text/plain"), Option.none(), Option.some(new Unsigned64(1000))));
        if (opened.isError()) {
            if (opened.error().tag() == 2) return new Unsigned64(10);
            throw new IllegalStateException("unexpected HTTP open error");
        }
        try (var upload = opened.value()) {
            if (which == 1) return new Unsigned64(1);
            LatentHttpStreaming.write(upload, new byte[]{100, 97, 116, 97}).value();
            var response = LatentHttpStreaming.finish(upload).value(); // consumes upload on any outcome
            try (var body = response.body()) {
                if (which == 2) return new Unsigned64(2);
                long count = 0;
                for (;;) {
                    var next = LatentHttpStreaming.read(body, 4L).value();
                    if (!next.isSome()) break;
                    try (var chunk = next.value()) {
                        if (which == 3) body.close(); // chunk remains an independent charged owner
                        count += LatentHttpStreaming.chunkBytes(chunk).value().length;
                        if (which == 3) return new Unsigned64(count);
                    }
                }
                LatentHttpStreaming.trailers(body).value();
                var again = LatentHttpStreaming.trailers(body);
                if (!again.isError() || again.error().tag() != 13) throw new IllegalStateException("trailers accepted twice");
                return new Unsigned64(count);
            }
        }
    }
}
